File: Medium/95CollidingAsteroids.js
function collidingAsteroids(asteroids) {
  const resultStack = [];

  for (const asteroid of asteroids) {
    if (asteroid > 0) {
      resultStack.push(asteroid);
      continue;
    }

    while (true) {
      if (resultStack.length === 0 || resultStack[resultStack.length - 1] < 0) {
        resultStack.push(asteroid);
        break;
      }

      const asteroidSize = Math.abs(asteroid);

      if (resultStack[resultStack.length - 1] > asteroidSize) {
        break;
      }

      if (resultStack[resultStack.length - 1] === asteroidSize) {
        resultStack.pop();
        break;
      }

      resultStack.pop();
    }
  }

  return resultStack;
}

File: Medium/95CollidingAsteroids.py
def collidingAsteroids(asteroids):
    resultStack = []
    for astroid in asteroids:
        if astroid > 0:
            resultStack.append(astroid)
            continue

        while True:
            if len(resultStack) == 0 or resultStack[-1] < 0:
                resultStack.append(astroid)
                break
            astroidSize = abs(astroid)
            if resultStack[-1] > astroidSize:
                break
            if resultStack[-1] == astroidSize:
                resultStack.pop()
                break
            resultStack.pop()
    return resultStack
